### ISSUE TITLE
Modify a process instance by activating elements inside a given flow scope

### ISSUE THREAD

**Issue by saig0** [camunda/zeebe#9646]
## Description

We want to allow the activation of elements inside a given flow scope instance (i.e. ancestor selection). The flow scope instance is selected by its element instance key. The selected instance could be the flow scope of the element or one of its upper scopes. 

By default, the new element instance is created within an existing flow scope instance (#9640 / #9642). Or, new flow scope instances are created if they aren't active yet (#9643).

But we need the additional option to select the ancestor scope instance:
* if multiple instances of the flow scope are active 
* if the element instance should be created in a new instance of a flow scope

If the selected ancestor scope is the direct flow scope then the new element instance is created within this scope instance. 

For example:
![image](https://user-images.githubusercontent.com/4305769/176424284-4cc30996-a645-499a-9a44-61a1800248e3.png)

If the selected ancestor scope is not the direct flow scope but an upper one (or the process instance itself) then new flow scope instances are created. Create a new flow scope instance for each flow scope between the selected ancestor and the given element. Create a new element instance of the given element within the created flow scope instance.   

For example:
![image](https://user-images.githubusercontent.com/4305769/176425826-ddf4f1d5-4373-43b0-914e-234c8f71b4a1.png)

**What needs to be done?**
* for each activate instruction:
  * if the selected ancestor flow scope instance is the direct flow scope then create the element instance as in #9642 
  * otherwise, create new flow scope instances under the selected ancestor as in #9643  

**These issues should be completed once ancestor selection is available**
* #9984
* #9990
* #11031
* #11032

**Out of scope**
* active an element inside a multi-instance

**Blocked by**
- #9643

**Comment by huygur** [comment]
Hi, I have some questions, most of them I think I already know the answer but I still would like to confirm:

1-` The selected instance could be the flow scope of the element or one of its upper scopes.` —> by selected instance, do you mean ‘ancestor instance key’ ?
2- `If the selected ancestor scope is not the direct flow scope but an upper one (or the process instance itself) then new flow scope instances are created. Create a new flow scope instance for each flow scope between the selected ancestor and the given element.` —> just to validate: it won’t matter how many instances the between elements have, right? Even if there was only 1 running instance in all the scopes in between, we still won’t be reusing them, instead will be creating new instances? I hope we have a clear way to communicate this in the UI 🤔 (@aleksander-dytko do we already have some designs for this ? )
3- if we don’t provide the ancestor instance key, will it behave as it is behaving now, or will it throw an exception? I am confused by this, because in the beginning of the issue it says` we need an additional option to select ancestor scope instance`, but in the end there is a task like this : `adjust the rejection message if no flow scope is selected (see Reject the modification command if multiple flow scope instances are active  #10327 (comment))`. Or is this just the case for multiple flow scopes?
4- what happens if I run multiple commands at once, with first command cancelling a sub process, and second command creating a new instance of a flow node within this subprocess providing the instance key of the already canceled sub process. (example: 
Sub Process
__2nd sub process
____Task B
 
command: [cancel all running instances in the sub process, create a new instance on Task B using the existing Sub Process instance key as ancestor key]

5- if there is only one running instance for each of the elements in the hierarchy below:
Sub Process
__2nd sub process
____Task B

what do I need to do in order to create a new instance on Task B, along with a new instance of all the parents? I assume if I don't provide any ancestor key, it will only create an instance for Task B and append it on the existing parent scope of Task B. should I be giving parent instance key as the ancestor id?

**Comment by aleksander-dytko** [comment]
@huygur 
>@aleksander-dytko do we already have some designs for this ?

Not yet - the assumption here is that Zeebe will start first on this, along with design and after that Operate.

**Comment by korthout** [comment]
@huygur Thanks for your input. Most of your assumptions are correct, except for the last one. Please have a look at my answers to your questions:

>1- `The selected instance could be the flow scope of the element or one of its upper scopes.` —> by selected instance, do you mean ‘ancestor instance key’ ?

ℹ️ By selected instance is meant: the instance referred to by the ancestor instance key. The point of that sentence is specifically to point out that the ancestor instance key can refer to the direct flow scope of the element to activate, or an indirect flow scope (an upper scope) of the element to activate (e.g. the direct flow scope of the direct flow scope, etc).

>2- it won’t matter how many instances the between elements have, right? Even if there was only 1 running instance in all the scopes in between, we still won’t be reusing them, instead will be creating new instances?

👍 Yes your assumption is correct. Any existing in-between instances are left untouched, and new instances are created that have the ancestor scope as their (direct/indirect) flow scope.

>3- if we don’t provide the ancestor instance key, will it behave as it is behaving now, or will it throw an exception?

ℹ️ Everything works like before when there is only a single instance.

ℹ️ It will behave as before, with a minor difference. If an activate instruction targets an element instance with a flow scope that has multiple active instances but no ancestor instance key is provided, then the rejection message will now be slightly different. Currently, it just says that this is rejected, but it does not say how to fix it (i.e. by providing an ancestor instance key) because ancestor selection was not yet implemented. We'll simply adjust the error message to be more helpful to users and understand that they can use the ancestor instance key to resolve this problem. See [this comment](https://github.com/camunda/zeebe/pull/10327#discussion_r969359329). 

>4- what happens if I run multiple commands at once

ℹ️ Zeebe will always process each command individually and either accept or reject the command. It then processes the next command. 

ℹ️ Having said that, the modify command can contain multiple activate and/or terminate instructions. We always execute the activations first, followed by the terminations. In your example (cancel a subprocess, activate an element inside the subprocess (possibly with flow scopes in between), we'll first activate the element inside the subprocess and then terminate the subprocess. The result is that the subprocess (including the activated element that has the subprocess as flow scope) has been terminated. 

>5- what do I need to do in order to create a new instance on Task B, along with a new instance of all the parents? 

ℹ️ You'll need to pass the process instance key as the ancestor instance key. This will create a new instance of the sub process (with the process instance itself as flow scope), a new instance of the 2nd sub process, and a new instance of the task. 

>I assume if I don't provide any ancestor key, it will only create an instance for Task B and append it on the existing parent scope of Task B.

❌ That's not correct. If you don't provide any ancestor key, the command will be rejected with the message (note the last sentence which will be a new addition to this rejection message:
```
Expected to modify instance of process 'x' but it contains one or more activate instructions for an element that has a flow scope with more than one active instances. Can't decide in which instance of the flow scope the element should be activated. **Please specify an ancestor element instance key for this activate instruction.**
```

Let me know, if you have any further questions about the behavior.

❓ @huygur @ralfpuchert With the additional information, please let me know if you see any problems with this behavior from Operate's perspective. Does it work as you would expect?

**Comment by huygur** [comment]
@korthout thank you for the answers! some of them are clear to me now, but some of them raised more questions 😄 

one I already asked here https://camunda.slack.com/archives/C0359JZEUV8/p1667919871981709 , as we might need more input on the topic, maybe there are some necessary changes in the UI.

 > ❌ That's not correct. If you don't provide any ancestor key, the command will be rejected with the message (note the last sentence which will be a new addition to this rejection message:

actually in my example there were only one running instances for each of the nested flow nodes:

Main Process
_Sub Process (1 running instance)
__2nd sub process (1 running instance)
____Task B (1 running instance)

I would expect that in this case, if I don't provide any ancestor key while starting a new instance on Task B, it will be like this, no?:

_Sub Process (1 running instance)
__2nd sub process (1 running instance)
____Task B (2 running instances)

**Comment by korthout** [comment]
Oops 😅 You're absolutely right @huygur! Sorry about the confusion.
